src/librustc_mir/transform/instrument_coverage.rs

```diff
@@ -0,0 +1,100 @@
+use crate::transform::{MirPass, MirSource};
+use rustc_index::vec::Idx;
+use rustc_middle::mir::interpret::Scalar;
+use rustc_middle::mir::*;
+use rustc_middle::mir::{Local, LocalDecl};
+use rustc_middle::ty;
+use rustc_middle::ty::Ty;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::def_id::DefId;
+use rustc_span::Span;
+
+pub struct InstrumentCoverage;
+
+/**
+ * Inserts call to count_code_region() as a placeholder to be replaced during code generation with
+ * the intrinsic llvm.instrprof.increment.
+ */
+
+// FIXME(richkadel): As a first step, counters are only injected at the top of each function.
+// The complete solution will inject counters at each conditional code branch.
+
+impl<'tcx> MirPass<'tcx> for InstrumentCoverage {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, src: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        if tcx.sess.opts.debugging_opts.instrument_coverage {
+            if let Some(callee_fn_def_id) = tcx.lang_items().count_code_region_fn() {
+                debug!("instrumenting {:?}", src.def_id());
+                instrument_coverage(tcx, callee_fn_def_id, body);
+            }
+        }
+    }
+}
+
+pub fn instrument_coverage<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    callee_fn_def_id: DefId,
+    body: &mut Body<'tcx>,
+) {
+    let span = body.span.shrink_to_lo();
+
+    let ret_ty = tcx.fn_sig(callee_fn_def_id).output();
+    let ret_ty = ret_ty.no_bound_vars().unwrap();
+    let substs = tcx.mk_substs(::std::iter::once(ty::subst::GenericArg::from(ret_ty)));
+
+    let count_code_region_fn: Operand<'_> =
+        Operand::function_handle(tcx, callee_fn_def_id, substs, span);
+
+    let index = const_int_operand(tcx, span.clone(), tcx.types.u32, 0);
+
+    let args = vec![index];
+
+    let source_info = SourceInfo { span: span, scope: OUTERMOST_SOURCE_SCOPE };
+
+    let new_block = START_BLOCK + body.basic_blocks().len();
+
+    let next_local = body.local_decls.len();
+    let new_temp = Local::new(next_local);
```

Comment:
You might want to make use of [`rustc_mir::util::patch::MirPatch`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/util/patch/struct.MirPatch.html). It would probably simplify this code a fair bit.

Reply:
Sounds good. I'll take another look at that. I referenced MirPatch when figuring out how to make that change and at first it was hard to tell how to insert a new start block. I think I have a better understanding now so I'll try to do that.

Reply:
Done